[ACIX-470] Create a top-level invoke task for running all gitlabci linters

Ishirui opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?

Building upon the work of #37417, continue refactoring the set of invoke tasks used for linting gitlabci-related in files in the agent repo. Notable refactoring changes include:
- Moving a lot of the task logic to `tasks/libs/linter` from the main `linter.py` file, leaving only the absolutely necessary parts in `linter.py`
- Unify the way tasks run on config diffs/full config objects/lists of jobs - before, each task had very different interfaces which was quite confusing. Now they can all be called in the same way and use similarly-built main inputs.
- Create a more robust system for handling and "bubbling up" linting failures to the top-level task, which allows each task to handle displaying failures as it prefers.

This PR also adds a new top-level task, `full-gitlab-ci`, that runs all the gitlabci-related linters together, sharing a preconfigured configuration object which takes a while to compute.
This new task is now run both in CI and as a pre-push hook, which increases the safety of gitlabci modifications.

### Motivation

We had a few linters for gitlabci that were only run manually, which is wasteful - we should run them in CI and/or as a pre-push. The issue is that a couple of these require a "fully-resolved" gitlabci configuration file as input, and computing that takes a while. Therefore it is critical that we do not compute this configuration file multiple times, which is easier if we use one single task grouping all the "sublinters".

### Describe how you validated your changes
<!--
Validate your changes before merge, ensuring that:
* Your PR is tested by static / unit / integrations / e2e tests
* Your PR description details which e2e tests cover your changes, if any
* The PR description contains details of how you validated your changes. If you validated changes manually and not through automated tests, add context on why automated tests did not fit your changes validation.

If you want additional validation by a second person, you can ask reviewers to do it. Describe how to set up an environment for manual tests in the PR description. Manual validation is expected to happen on every commit before merge.

Any manual validation step should then map to an automated test. Manual validation should not substitute automation, minus exceptions not supported by test tooling yet.
-->

Running all linters, also with failing examples + unit tests and in CI.

### Possible Drawbacks / Trade-offs

Performance: we are now running more linters, especially as a pre-push, which makes it quite a bit slower. This might make changes to gitlabci files more frustrating to make as the pre-push hook that is run when modifying them is quite a bit slower than it was before.

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

Ishirui committed:
refactor(linter): Transform `get_gitlab_ci_lintable_jobs` into `extract_gitlab_ci_jobs`

We remove all the file loading functionality from this method and move it into the tasks themselves. This allows for better separation of concerns and also allows the new method to have a more clear purpose.

Ishirui committed:
feat(linter): Make `gitlab_change_paths` use the diff instead of the full config

Since we will end up precomputing the diff anyway, this will make this check faster (less jobs to look through)

Ishirui committed:
refactor(linter): Move task logic to `libs/`

This also allows use to use non-task functions to accomplish linting on already-loaded config objects, which will be useful for creating the top-level linting task.

Ishirui committed:
feat(linter): Use `load_or_generate` helpers to only generate config if needed

fix(gitlab): Properly check for `diff is None`

Ishirui committed:
feat(linter): Create top-level task combining all gitlabci-related linters

Ishirui committed:
feat(linter): Make `check_change_paths_gitlab_ci_config` handle job lists

This simplifies the logic of this function and also brings its interface in line with the rest of the linting methods, which take in lists of jobs instead of full config objects.

refactor(linter): Rename `check_change_paths_gitlab_ci_jobs` and `check_change_paths_gitlab_ci_config`

Make it clearer what the difference between the two is, and correctly say the second one handles jobs, not a full config

Ishirui committed:
feat(linter): Create unified `Exception` class for handling linting failures

The idea is to use Exceptions to make linting failures go back to top-level, that way each top-level task can customize how it handles/displays linting errors, instead of having `print` statements buried in "plumbing" functions (like the ones in `libs/`

feat(linter): Use new exception-based failure reporting mechanism for all linting tasks

Ishirui committed:
feat(linter): Make all linting tasks able to run on the full config

When using the individual linting tasks we'll probably want to check all the jobs and not have to worry about the diff (it wouldn't really save much time anyway, generating the config or diff is what takes long)

Ishirui committed:
refactor(linter): Move all common task logic (config loading and failure printing) to its own function

Ishirui committed:
feat(linter): Add better error handling and logging to `full_gitlab_ci` + fix unit tests

Ishirui committed:
feat(ci): Run full gitlabci linters on pre-push and in CI

Ishirui commented on .pre-commit-config.yaml:
```diff
@@ -118,21 +118,13 @@ repos:
-      description: test the gitlab configuration on main
-      entry: 'dda inv linter.gitlab-ci -t main'
+      description: Lint and test GitLab CI configuration files
+      entry: "dda inv linter.full-gitlab-ci -t main"
```
This is still quite slow, even for a pre-push. Running it on the diff instead also does not help too much, as generating the full config/diff is actually one of the more time-consuming parts. Maybe we should disable some checks from the "full" linter suite when running as pre-push and only run them in CI (I think shellcheck is quite slow for example ?)

I will do some proper profiling to see which step(s) take the most time.

CelianR replied:
Another thing, IMO we should have `--fail-fast` here since it is a hook

Ishirui committed:
feat(linter): Allow passing in a jobowners object instead of a filepath to `check_owners_gitlab_ci_jobs`

Ishirui committed:
refactor(linter): Reformat docstrings using `docformatter`

Ishirui commented on tasks/linter.py:
```diff
@@ -236,101 +238,208 @@ def github_actions_shellcheck(
+    shellcheck_only_errors: bool = False,
+    path_jobowners: str = '.gitlab/JOBOWNERS',
+):
+    """
+    Top-level task for running all gitlabci-related linters
```
I wasn't aware of the convention 😓 Thanks for noticing ! Fixed.

Ishirui committed:
refactor(linter): Replace all calls to `sys.exit` with `raise Exit`

Ishirui committed:
refactor(linter): Improve OOP pattern for `GitlabLintFailure`

Ishirui committed:
fix(linter): Improve error printing for `github_actions_shellcheck`

Ishirui committed:
feat(linter): Add section for pretty-printing and exiting with an error code at the end of `full_gitlab_ci`

Ishirui committed:
feat(linter): Implement `gitlabci_sublinter` decorator

This allows us to centralize config options for linters near the actual method definition, and automatically register those linters for usage in `full_gitlab_ci`

Ishirui committed:
fix: typo in error message

Ishirui committed:
refactor(linter): Use `gitlabci_lint_task_template` for `gitlab_ci_shellcheck` and `full_gitlab_ci`

Ishirui committed:
feat(linter): Add `allow_fail` and `run_on_prepush` options to `gitlabci_sublinter` decorator

Ishirui committed:
refactor(linter): Use IGNORE failure level where appropriate

Ishirui committed:
feat(linter): Add verbosity argument

CelianR commented on tasks/libs/linter/gitlab_exceptions.py:
```diff
@@ -0,0 +1,151 @@
+class FailureLevel(int, Enum):
+    """Enum for different criticalities of gitlabci linting failures."""
+
+    CRITICAL = 3  # Something went wrong while linting
+    ERROR = 2  # The linter found something wrong with the file being linted
```
Is there a real difference between both?

CelianR commented on tasks/libs/linter/gitlab_exceptions.py:
```diff
@@ -0,0 +1,151 @@
+        level_out = self.level.pretty_print()
+        if len(self.entry_points) > 1:
+            entry_point = ", ".join(self.entry_points)
+        elif len(self.entry_points) == 1:
+            entry_point = next(iter(self.entry_points))
```
I think it isn't necessary since `', '.join(['a']) == 'a'`

CelianR commented on tasks/libs/linter/gitlab.py:
```diff
@@ -1,22 +1,403 @@
+            # We never get here if the function raises an exception
+            print(f"[{color_message('OK', Color.GREEN)}] {success_message}")
+
+        ALL_GITLABCI_SUBLINTERS.append(wrapper)
+        if run_on_prepush:
+            PREPUSH_GITLABCI_SUBLINTERS.append(wrapper)
+        return wrapper
```
Nice 👏

CelianR commented on tasks/linter.py:
```diff
@@ -236,109 +242,182 @@ def github_actions_shellcheck(
+            "jobowners": jobowners,
+        }
+
+        to_run = ALL_GITLABCI_SUBLINTERS if running_in_ci() else PREPUSH_GITLABCI_SUBLINTERS
```
NIT: I think we should have an option to run everything locally but which isn't enabled by default

Ishirui replied:
Or maybe even go further and add a seperate flag (`--pre-push-linters`) that would be set in the `.pre-commit-config.yaml`, this way the task runs everything by default when called manually ?

CelianR reviewed (approved):
Gg! Minor suggestions

CelianR commented on tasks/linter.py:
```diff
@@ -236,109 +238,230 @@ def github_actions_shellcheck(
+        else:
+            grouped_failure = MultiGitlabLintFailure(failures)
+
+        with gitlab_section('GitlabCI Linter failures'):
```
I think we have to do sections for each sub linter and also for the failure summary

CelianR replied:
Indeed it's interesting to see from which linter the error comes from (even if we can see this via the stack trace)

Ishirui committed:
feat(linter): Address review comments:
- Minor code changes
- Remove CRITICAL failure level
- Add `--pre-push-linters` flag
- Run pre-push linters with `--fail-fast`

Ishirui committed:
fix(linter): Don't run `shellcheck` in CI

The tool is not installed on our runners yet, so it always fails

Ishirui committed:
feat(linter): Track each sublinter's name and improve summarized failure output

Ishirui committed:
fix(.gitlab): Add new job to JOBOWNERS

Ishirui committed:
fix: formatting

CelianR commented on tasks/libs/linter/gitlab.py:
```diff
@@ -152,6 +158,12 @@ def shellcheck_gitlab_ci_jobs(
     Note:
         Will raise an Exit if any errors are found.
     """
+    # TODO[@agent-devx]: Remove this once we have shellcheck in CI
```
```suggestion
    # TODO(@agent-devx): Remove this once we have shellcheck in CI
```

Ishirui committed:
Update tasks/libs/linter/gitlab.py

Co-authored-by: Célian Raimbault <161456554+CelianR@users.noreply.github.com>
